src/posit/connect/resources.py

```diff
@@ -43,3 +46,254 @@ def update(self, *args, **kwargs):
+        self._ctx = ctx
+        self._parent = parent
+
+
+T_co = TypeVar("T_co", bound="Active", covariant=True)
+"""A covariant type variable that is bound to the `Active` class, meaning that `T_co` must be or derive from `Active`."""
+
+
+class ActiveSequence(ABC, Generic[T_co], Sequence[T_co]):
+    """
+    A sequence abstraction for any HTTP GET endpoint that returns a collection.
+
+    It lazily fetches data on demand, caches the results, and allows for standard sequence operations like indexing and slicing.
+
+    Parameters
+    ----------
+    cls : Type[T_co]
+        The class used to represent each item in the sequence.
+    ctx : Context
+        The context object that holds the HTTP session used for sending the GET request.
+    parent : Optional[Active], optional
+        An optional parent resource to establish a nested relationship, by default None.
+
+    Attributes
+    ----------
+    _cls : Type[T_co]
+        The class used to instantiate each item in the sequence.
+    _ctx : Context
+        The context containing the HTTP session used to interact with the API.
+    _parent : Optional[Active]
+        Optional parent resource for maintaining hierarchical relationships.
+    _cache : Optional[List[T_co]]
+        Cached list of items returned from the API. Set to None initially, and populated after the first request.
+
+    Abstract Properties
+    -------------------
+    _endpoint : str
+        The API endpoint URL for the HTTP GET request. Subclasses are required to implement this property.
+
+    Methods
+    -------
+    _data() -> List[T_co]
+        Fetch and cache the data from the API. This method sends a GET request to `_endpoint`, parses the
+        response as JSON, and instantiates each item using `cls`.
+
+    __getitem__(index) -> Union[T_co, Sequence[T_co]]
+        Retrieve an item or slice from the sequence. Indexing follows the standard Python sequence semantics.
+
+    __len__() -> int
+        Return the number of items in the sequence.
+
+    __str__() -> str
+        Return a string representation of the cached data.
+
+    __repr__() -> str
+        Return a detailed string representation of the cached data.
+
+    reload() -> ActiveSequence
+        Clear the cache and mark to reload the data from the API on the next operation.
+    """
+
+    def __init__(self, cls: Type[T_co], ctx: Context, parent: Optional[Active] = None):
+        super().__init__()
+        self._cls = cls
+        self._ctx = ctx
+        self._parent = parent
+        self._cache = None
+
+    @property
+    @abstractmethod
+    def _endpoint(self) -> str:
+        """
+        Abstract property to define the endpoint URL for the GET request.
+
+        Subclasses must implement this property to return the API endpoint URL that will
+        be queried to fetch the data.
+
+        Returns
+        -------
+        str
+            The API endpoint URL.
+        """
+        raise NotImplementedError()
+
+    @property
```

Comment:
Because this is doing more than a special retrieval, I'd argue for a true method over a property... Let's pkg authors know it's doing something and not just immediately returning a value

Reply:
By having a method, it also removes the possibility of setting the value